sdk/eventhub/Azure.Messaging.EventHubs/src/Core/ChannelEnumerableSubscription.cs

```diff
@@ -116,134 +121,84 @@ public async ValueTask DisposeAsync()
                                                                   [EnumeratorCancellation]CancellationToken cancellationToken)
         {
             T result;
-            int delayMilliseconds;
 
-            var waitTime = maximumWaitTime?.TotalMilliseconds;
-            var timer = (waitTime.HasValue) ? Stopwatch.StartNew() : default;
+            var waitToken = cancellationToken;
+            var waitCancel = default(CancellationTokenSource);
 
             while (!cancellationToken.IsCancellationRequested)
             {
-                // Attempt to read events from the channel.
-
                 if (reader.TryRead(out result))
                 {
                     yield return result;
-                    timer?.Restart();
-                }
-                else if ((waitTime.HasValue) && (timer.ElapsedMilliseconds > waitTime.Value))
-                {
-                    // If there was no event and the wait time has expired, emit an empty event to return control to
-                    // the calling iterator.
-
-                    yield return default;
-                    timer.Restart();
                 }
                 else if (reader.Completion.IsCompleted)
                 {
                     // If the channel was marked as final, then await the completion task to surface any exceptions.
 
-                    try
-                    {
-                        await reader.Completion.ConfigureAwait(false);
-                    }
-                    catch (TaskCanceledException)
-                    {
-                        // This is an expected case when the cancellation token was
-                        // triggered during an operation; no action is needed.
-                    }
-
+                    await reader.Completion.ConfigureAwait(false);
                     break;
                 }
                 else
                 {
-                    // TODO (P1//squire): Determine a better approach to prevent a tight loop that consumes resources or
-                    // has excessive allocations.  (see: https://github.com/Azure/azure-sdk-for-net/issues/7094)
-
                     try
                     {
-                        delayMilliseconds = 50;
-
-                        if (waitTime.HasValue)
+                        if (maximumWaitTime.HasValue)
                         {
-                            delayMilliseconds = (int)Math.Min(delayMilliseconds, (waitTime.Value - timer.ElapsedMilliseconds));
+                            waitCancel = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
```

Comment:
This is potentially an allocation per iteration, can we move the initialization out of the loop?

Reply:
Unless you know a trick to reset a token source that has been canceled, I don't think we can hoist it.   I did make some tweaks to try and extend the token if we didn't hit the wait timeout in order to keep from having to reallocate each iteration.    Thoughts?

Reply:
Right, that's what I've meant, calling CancelAfter while the token is "alive" and recreating it after it expired.